### PR TITLE
Run scalar integer index normalization at trace time

### LINKED ISSUE #2878: _index_to_gather produces DeviceArrays for int / onp.ndarray inputs
The `_index_to_gather` function and some subfunctions, like `_normalize_index`, perform operations with `lax` functions even when dealing with integer or onp.ndarray inputs. That's unfortunate because it means, for example, doing scalar integer operations on the default accelerator device! We should probably revise.

### PR DESCRIPTION
**Pull request by lgeiger**
This PR speeds up scalar integer indexing outside of jit by doing index computations in plain Python.

I am seeing a ~3x speedup of following toy example on CPU without negatively affecting trace time:
```python
from jax import numpy as jnp

a = jnp.arange(100)
a[5]
%timeit a[6].block_until_ready()
```

**before**:
```
{ lambda ; a:i32[100]. let
    b:bool[] = lt 5 0
    c:i32[] = add 5 100
    d:i32[] = select_n b 5 c
    e:i32[] = convert_element_type[new_dtype=int32 weak_type=False] d
    f:i32[1] = broadcast_in_dim[broadcast_dimensions=() shape=(1,)] e
    g:i32[] = gather[
      dimension_numbers=GatherDimensionNumbers(offset_dims=(), collapsed_slice_dims=(0,), start_index_map=(0,))
      fill_value=None
      indices_are_sorted=True
      mode=GatherScatterMode.PROMISE_IN_BOUNDS
      slice_sizes=(1,)
      unique_indices=True
    ] a f
  in (g,) }
```

**after:**
```
{ lambda a:i32[1]; b:i32[100]. let
    c:i32[] = gather[
      dimension_numbers=GatherDimensionNumbers(offset_dims=(), collapsed_slice_dims=(0,), start_index_map=(0,))
      fill_value=None
      indices_are_sorted=True
      mode=GatherScatterMode.PROMISE_IN_BOUNDS
      slice_sizes=(1,)
      unique_indices=True
    ] b a
  in (c,) }
```

#10393 was reverted since it included many changes that had unintended side effects which made it hard to review and ended up breaking user code, sorry about that. This PR is a stripped down version that tries to limit the the diff to not reducing code readability.

Fixes #2878

### CODE REVIEW

**Comment by jakevdp** [comment]
With all these kinds of speedups, a question that comes to mind is this: should `a.__getitem__(i)` and `jit(a.__getitem__)(i)` perform different operations? If you're detecting when `i` is a static value, then effectively what you're doing is adding a special code-path that only executes when the function is not jit-compiled.

I don't doubt this may be faster, but I'm not certain a 3x speedup of an already fast operation is worth the added complexity in implementation and in understanding how a particular operation is maps to core operations.

**Comment by jakevdp** [comment]
Since I just saw this issue and it's at the top of my mind, here's an example of an optimization we did where a different code-path is chosen for static integer inputs, and it leads to a confusing divergence of behavior in a corner case: #10692

The more such optimizations the add, the more potential there is for these kinds of sharp corners.

**Comment by YouJiacheng** [comment]
I think it is a rare case to perform millions of scalar indexing without `jit`.
